cpp/tests/groupby/structs_tests.cpp

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (c) 2019-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <tests/groupby/groupby_test_util.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/iterator_utilities.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include "cudf/aggregation.hpp"
+#include "cudf/types.hpp"
+
+using namespace cudf::test::iterators;
+
+namespace cudf {
+namespace test {
+
+template <typename V>
+struct groupby_structs_test : public cudf::test::BaseFixture {
+};
+
+TYPED_TEST_CASE(groupby_structs_test, cudf::test::FixedWidthTypes);
+
+using V       = int32_t;  // Type of Aggregation Column.
+using M0      = int32_t;  // Type of STRUCT's first (i.e. 0th) member.
+using R       = cudf::detail::target_type_t<V, aggregation::SUM>;  // Type of aggregation result.
+using offsets = std::vector<cudf::offset_type>;
+using strings = strings_column_wrapper;
+using structs = structs_column_wrapper;
+
+template <typename T>
+using fwcw = fixed_width_column_wrapper<T>;
+
+template <typename T>
+using lcw = lists_column_wrapper<T>;
+
+namespace {
+static constexpr auto null = -1;  // Signifies null value.
+
+// Checking with a single aggregation, and aggregation column.
+// This test is orthogonal to the aggregation type; it focuses on testing the grouping
+// with STRUCT keys.
+auto sum_agg() { return cudf::make_sum_aggregation<groupby_aggregation>(); }
+
+#ifndef NDEBUG
+void print_agg_results(column_view const& keys, column_view const& vals)
+{
+  auto requests = std::vector<groupby::aggregation_request>{};
+  requests.push_back(groupby::aggregation_request{});
+  requests.back().values = vals;
+  requests.back().aggregations.push_back(sum_agg());
+  requests.back().aggregations.push_back(
+    cudf::make_nth_element_aggregation<groupby_aggregation>(0));
+
+  auto gby    = groupby::groupby{table_view({keys}), null_policy::INCLUDE, sorted::NO, {}, {}};
+  auto result = gby.aggregate(requests);
+  std::cout << "Results: Keys: " << std::endl;
+  print(result.first->get_column(0).view());
+  std::cout << "Results: Values: " << std::endl;
+  print(result.second.front().results[0]->view());
+}
+#define PRINT_AGG_RESULTS(K, V) print_agg_results((K), (V))
+#else
+#define PRINT_AGG_RESULTS(K, V)
+#endif
```

Comment:
I don't think always printing debug in debug mode is accepted. Maybe:
```suggestion
constexpr bool print_debug{false};
void print_agg_results(column_view const& keys, column_view const& vals)
{
if constexpr(print_debug) {

........
}
} // end void print_agg_results
```

Reply:
Agreed. I've moved the printing behind a different flag, instead of using `NDEBUG`.